arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/context/RaiseAccumulateContext.kt

```diff
@@ -8,113 +8,218 @@ package arrow.core.raise.context
 import arrow.core.Either
 import arrow.core.EitherNel
 import arrow.core.NonEmptyList
+import arrow.core.NonEmptySet
+import arrow.core.getOrElse
+import arrow.core.raise.Accumulate
 import arrow.core.raise.ExperimentalRaiseAccumulateApi
 import arrow.core.raise.RaiseAccumulate.Value
 import arrow.core.raise.RaiseDSL
+import arrow.core.raise.RaiseNel
+import arrow.core.raise.accumulating
+import arrow.core.raise.ensureNotNullOrAccumulate
+import arrow.core.raise.ensureOrAccumulate
 import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.InvocationKind
+import kotlin.contracts.InvocationKind.AT_LEAST_ONCE
 import kotlin.contracts.InvocationKind.AT_MOST_ONCE
 import kotlin.contracts.contract
+import kotlin.experimental.ExperimentalTypeInference
+import kotlin.jvm.JvmMultifileClass
+import kotlin.jvm.JvmName
 import arrow.core.raise.accumulate as accumulateExt
 import arrow.core.raise.mapOrAccumulate as mapOrAccumulateExt
 import arrow.core.raise.mapValuesOrAccumulate as mapValuesOrAccumulateExt
 import arrow.core.raise.zipOrAccumulate as zipOrAccumulateExt
-import kotlin.experimental.ExperimentalTypeInference
-import kotlin.jvm.JvmMultifileClass
-import kotlin.jvm.JvmName
 
+@Deprecated("RaiseAccumulate has been split into Raise and Accumulate. Use the new APIs instead.")
 public typealias RaiseAccumulate<A> = arrow.core.raise.RaiseAccumulate<A>
 
+// without the (this, this) we get an internal compiler error. May be fixed in 2.2.20?
 context(raise: Raise<NonEmptyList<Error>>)
 @ExperimentalRaiseAccumulateApi @RaiseDSL public inline fun <Error, A> accumulate(
-  block: context(RaiseAccumulate<Error>) () -> A
-): A = raise.accumulateExt(block)
+  block: context(Raise<Error>, Accumulate<Error>) () -> A
```

Comment:
I think this paints a wrong picture of what is going on. As far as I understand, you're not supposed to use `Raise` and `Accumulate` separately, they should always come in tandem. But by making them separate context parameters, you risk resolution choosing different ones.

Reply:
Fair enough yes! I'll separate the context stuff into another PR, likely with way more explanation and so we may discuss the design there.